### PR TITLE
feat(SchedulePagination): add pagination to schedule component

### PR DESCRIPTION
**Pull request by Ojaghi1991**
Please ensure your pull request adheres to the following guidelines:

- [ ] Use the following format: `* [owner/repo](link)`
- [ ] Link additions should be added to the bottom of the relevant category.
- [ ] New categories or improvements to the existing categorization are welcome.
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] Sort by alphabetical order

Thanks for contributing!

### CODE REVIEW

File: components/molecules/charts/schedule/columnHelper.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
+import {
```
Comment (iwarner) [inline components/molecules/charts/schedule/columnHelper.js]: @Ojaghi1991 Please remember the consistency - Doc Blocks - // React - style about UI and third party libraries above

Space code logicially etc

One export per file must be default

Move other large blocks to helper folder and seperate files also please

File: components/molecules/charts/schedule/helper.js [hunk 2]
```diff
@@ -0,0 +1,292 @@
+import React from 'react'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper.js]: @Ojaghi1991 Seperate into indivudal files - this helps with code bloat if we dont need to use all the methods in here - easier for testing and code maintenance also

camelCase we have a util for already - dont need lodash for that i thhink

Also prefer to remove lodash if we can from all files

File: components/molecules/charts/schedule/schedule.js [hunk 3]
```diff
@@ -3,26 +3,16 @@
...
-import Tooltip from '../../../atoms/tooltip/tooltip'
-import THEME_SIZE from '../../../constants/size'
 import Pagination from '../../../molecules/pagination/pagination'
-import Row from '../../../atoms/grid/Row'
+
+import { columns, ScheduleToolbar } from './columnHelper'
+import { prepareHiddenColumn, prepareScheduleRows, generateFilterDate } from './helper'
 
 const DATE_TYPE = {
   YEAR: 'year',
```
Comment (iwarner) [inline components/molecules/charts/schedule/schedule.js]: @Ojaghi1991 Alpha

File: components/molecules/charts/schedule/schedule.js [hunk 4]
```diff
@@ -3,26 +3,16 @@
...
 import Table from '../../../molecules/table/table'
-import Text from '../../../atoms/text/text'
-import THEME_ALIGN from '../../../constants/align'
-import Tooltip from '../../../atoms/tooltip/tooltip'
-import THEME_SIZE from '../../../constants/size'
 import Pagination from '../../../molecules/pagination/pagination'
-import Row from '../../../atoms/grid/Row'
+
+import { columns, ScheduleToolbar } from './columnHelper'
+import { prepareHiddenColumn, prepareScheduleRows, generateFilterDate } from './helper'
```
Comment (iwarner) [inline components/molecules/charts/schedule/schedule.js]: @Ojaghi1991 Alpha

File: components/molecules/charts/schedule/schedule.js [hunk 5]
```diff
@@ -3,26 +3,16 @@
  */
 
 // React
-import React, { useState } from 'react'
-import { array, func, oneOf } from 'prop-types'
-
-// Style
-import styled from 'styled-components'
+import React, { useEffect, useState } from 'react'
+import { array, func, string, oneOf } from 'prop-types'
```
Comment (iwarner) [inline components/molecules/charts/schedule/schedule.js]: @Ojaghi1991 Allpha

File: components/molecules/charts/schedule/schedule.js [hunk 6]
```diff
@@ -31,162 +21,69 @@ const DATE_TYPE = {
...
-  handleClick,
-  handleRowClick,
+  dataSource,
+  events,
+  eventTimeSplitting,
+  flag,
+  hiddenColumn,
   initialMode,
+  title,
+  yearRange,
```
Comment (iwarner) [inline components/molecules/charts/schedule/schedule.js]: @Ojaghi1991 Alpha

**Review by iwarner — CHANGES_REQUESTED** [review]
Please check ASAP - PRs become a priority to fix when reviwed thanks

File: components/molecules/charts/schedule/helper/columns.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+import {
+  endOfMonth,
+  endOfDay,
+  endOfWeek,
+  startOfMonth,
+  startOfWeek,
+  startOfDay,
+  setMonth,
+  setDay
+} from 'date-fns'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js:21]: @Ojaghi1991 Space here

File: components/molecules/charts/schedule/helper/columns.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+// IUI
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js]: @Ojaghi1991 Name UI not IUI

File: components/molecules/charts/schedule/helper/columns.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+const DATE_TYPE = {
+  DAY: 'day',
+  MONTH: 'month',
+  YEAR: 'year',
+  WEEK: 'week'
+}
+
+const columnPattern = {
+  day: [...Array(12).keys()],
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js:29]: @Ojaghi1991 If constants should be CAPTIAL

File: components/molecules/charts/schedule/helper/columns.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  startOfWeek,
+  startOfDay,
+  setMonth,
+  setDay
+} from 'date-fns'
+// IUI
+import ColumnPagination from './columnPagination'
+import Tooltip from '../../../../atoms/tooltip/tooltip'
+
+const DATE_TYPE = {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js]: @Ojaghi1991 We have a  global contstant folder maybe this belongs there?

File: components/molecules/charts/schedule/helper/columns.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  month: ['Week1', 'Week2', 'Week3', 'Week4'],
+  year: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
+  week: ['Sat', 'Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri']
+}
+
+const formatCell = (handleClick, month, row, otherData) => {
+  return row[month] && row[month][0] ? (
+    row[month][2] ? (
+      <Tooltip content={row[month][2]}>
+        <Wrapper
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js]: @Ojaghi1991 Wrapper is duped below - keep DRY and sepearate into file and component of itself

File: components/molecules/charts/schedule/helper/columns.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+export default (handleClick, options) => {
+  const {
+    mode = DATE_TYPE.MONTH,
+    setCurrentDate,
+    currentDate,
+    hiddenColumn,
+    currentDataSource,
+    onTitleFormatter
+  } = options
+  if (!currentDataSource.length) return []
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js]: @Ojaghi1991 Always use brackets with If () { } and add space above for readabillity

File: components/molecules/charts/schedule/helper/columns.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
+/**
+ * Column
+ */
+
+// React
+import React from 'react'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js:6]: @Ojaghi1991 Space below

// Style
...

No Proptypes on this component - there should be?

File: components/molecules/charts/schedule/helper/filterDateCriteria.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { getWeek, getDate } from 'date-fns'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/filterDateCriteria.js:9]: @Ojaghi1991 Docblocks !!

File: components/molecules/charts/schedule/helper/generateFilterDate.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+import {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/generateFilterDate.js:9]: @Ojaghi1991 Docblock

File: components/molecules/charts/schedule/helper/prepareScheduleRows.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
+import { groupBy } from 'lodash'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/prepareScheduleRows.js:8]: @Ojaghi1991 Dockblock

File: components/molecules/charts/schedule/schedule.js [hunk 17]
```diff
@@ -3,190 +3,90 @@
...
-import Row from '../../../atoms/grid/Row'
+
+import columns from './helper/columns'
+import generateFilterDate from './helper/generateFilterDate'
+import prepareHiddenColumn from './helper/prepareHiddenColumn'
+import prepareScheduleRows from './helper/prepareScheduleRows'
 
 const DATE_TYPE = {
-  YEAR: 'year',
+  DAY: 'day',
```
Comment (iwarner) [inline components/molecules/charts/schedule/schedule.js]: @Ojaghi1991 repeated in other file hence suggestion to move to global

**Review by iwarner — CHANGES_REQUESTED** [review]
@Ojaghi1991 Still lots of semantics with consistency of code style please review again

File: components/molecules/charts/schedule/helper/columns.js (hunk 13 above)
Comment (Ojaghi1991) [reply components/molecules/charts/schedule/helper/columns.js:6]: Columns is not Component, it's helper function to generate column

File: components/constants/dateType.js [hunk 18]
```diff
@@ -0,0 +1,12 @@
+/**
+ * DATE TYPE
```
Comment (iwarner) [inline components/constants/dateType.js]: @Ojaghi1991 Put path is as others :)

File: components/molecules/charts/schedule/helper/columnPagination.js [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+import Row from '../../../../atoms/grid/Row'
+
+const args = {
+  align: THEME_ALIGN.Start,
+  size: THEME_SIZE.MD
+}
+
+const calculateCurrent = (currentDate, mode, status = 'current') => {
+  const pCurrentDate = Date.parse(currentDate)
+  switch (mode) {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columnPagination.js:36]: @Ojaghi1991 On Switch best practices is to always have a default right. Choose the most common case for the component and make that the default - probably year?

Also the inner conditionals should be switch also - I think switch is always faster than if else

File: components/molecules/charts/schedule/helper/columnPagination.js [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+        )}`
+      } else if (status === 'next') {
+        return addDays(pCurrentDate, 1)
+      } else if (status === 'prev') {
+        return subDays(pCurrentDate, 1)
+      }
+  }
+}
+
+const handleChangeDate = (setCurrentDate, options) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columnPagination.js]: @Ojaghi1991 Did we say we were always going to pass objects to methods so 

const handleChangeDate = ({options, setCurrentDate})

I like this also as easy to ALPHA then :)

File: components/molecules/charts/schedule/helper/columnPagination.js [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+  const { currentDate, mode, status } = options
+  const current = calculateCurrent(currentDate, mode, status)
+  setCurrentDate(current.toISOString())
+}
+
+export default ({ setCurrentDate, mode, currentDate }) => {
+  const current = calculateCurrent(currentDate, mode, 'current')
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columnPagination.js]: @Ojaghi1991 Space below this and next logical seems right

File: components/molecules/charts/schedule/helper/columnPagination.js [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+  }
+}
+
+const handleChangeDate = (setCurrentDate, options) => {
+  const { currentDate, mode, status } = options
+  const current = calculateCurrent(currentDate, mode, status)
+  setCurrentDate(current.toISOString())
+}
+
+export default ({ setCurrentDate, mode, currentDate }) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columnPagination.js]: @Ojaghi1991 Alpha

File: components/molecules/charts/schedule/helper/columnPagination.js [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+  const current = calculateCurrent(currentDate, mode, 'current')
+  return (
+    <Row justify={'center'}>
+      <ButtonToolbar {...args}>
+        <Button
+          onClick={() => handleChangeDate(setCurrentDate, { currentDate, mode, status: 'prev' })}
+        >
+          &lt;
+        </Button>
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columnPagination.js:107]: @Ojaghi1991 I like spaces between these blocks also - this is more optional but I think it helps with looking down the lines personally

File: components/molecules/charts/schedule/helper/columns.js [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+} from 'date-fns'
+
+// UI
+import ColumnPagination from './columnPagination'
+import Tooltip from '../../../../atoms/tooltip/tooltip'
+import DATE_TYPE from '../../../../constants/dateType'
+
+const COLUMN_PATTERN = {
+  day: [...Array(12).keys()],
+  month: ['Week1', 'Week2', 'Week3', 'Week4'],
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/columns.js]: @Ojaghi1991 There can sometimes be 5 weeks in a month - be careful with that, test this works in such occassions - also what day or the week does your week start

File: components/molecules/charts/schedule/helper/filterDateCriteria.js [hunk 25]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Components - Molecules - Charts - Schedule - Helper - Filter Date Criteria
+ */
+
+// Date FNS
+import { getWeek, getDate } from 'date-fns'
+
+export default (type, timingStart) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/filterDateCriteria.js]: @Ojaghi1991 Default here

File: components/molecules/charts/schedule/helper/generateFilterDate.js [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+  endOfWeek,
+  endOfMonth,
+  endOfDay,
+  startOfMonth,
+  startOfWeek,
+  startOfDay,
+  startOfYear
+} from 'date-fns'
+
+export default (status, date) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/generateFilterDate.js]: @Ojaghi1991 Default

File: components/molecules/charts/schedule/helper/prepareHiddenColumn.js [hunk 27]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Components - Molecules - Charts - Schedule - Helper - Hidden Column
+ */
+
+export default (hiddenColumn) => {
+  const result = {}
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/prepareHiddenColumn.js:9]: @Ojaghi1991 Nice spacing

File: components/molecules/charts/schedule/helper/prepareHiddenColumn.js [hunk 28]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Components - Molecules - Charts - Schedule - Helper - Hidden Column
+ */
+
+export default (hiddenColumn) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/prepareHiddenColumn.js]: @Ojaghi1991 I think proptypes can still work on non react methods - I have used them, prob need to test but lets see

File: components/molecules/charts/schedule/helper/prepareScheduleRows.js [hunk 29]
```diff
@@ -0,0 +1,148 @@
+/**
+ * Components - Molecules - Charts - Schedule - Helper - Prepare Schedule Rows
+ */
+
+// Lodash
+import { groupBy } from 'lodash'
+// Date FNS
+import { daysToWeeks, format } from 'date-fns'
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/prepareScheduleRows.js:11]: @Ojaghi1991 Space below

File: components/molecules/charts/schedule/helper/prepareScheduleRows.js [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+// UI
+import camelCase from '../../../../utils/camelCase/camelCase'
+
+const generateDate = (type, arr) => {
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/prepareScheduleRows.js:16]: @Ojaghi1991 default

File: components/molecules/charts/schedule/helper/scheduleToolbar.js [hunk 31]
```diff
@@ -0,0 +1,58 @@
+/**
+ * Components - Molecules - Charts - Schedule - Helper - Schedule Toolbar
+ */
+// React
```
Comment (iwarner) [inline components/molecules/charts/schedule/helper/scheduleToolbar.js:5]: @Ojaghi1991 Space above

**Review by iwarner — CHANGES_REQUESTED** [review]
@Ojaghi1991 Lots better just some more semantics

File: components/molecules/charts/schedule/helper/columns.js (hunk 24 above)
Comment (Ojaghi1991) [reply components/molecules/charts/schedule/helper/columns.js]: @iwarner handled with eventTimeSplitting props and listen to that and used  these methods 
`  endOfMonth,
  endOfDay,
  endOfWeek,
  startOfMonth,
  startOfWeek,
  startOfDay,
`

**Review by iwarner — APPROVED** [review]

**Pull request merged**